mllib/src/main/scala/org/apache/spark/ml/feature/Instance.scala

```diff
@@ -114,6 +133,65 @@ private[spark] object InstanceBlock {
   def blokify(instances: RDD[Instance], blockSize: Int): RDD[InstanceBlock] = {
     instances.mapPartitions(_.grouped(blockSize).map(InstanceBlock.fromInstances))
   }
+
+  def blokifyWithMaxMemoryUsage(
+      iterator: Iterator[Instance],
+      maxMemoryUsage: Long): Iterator[InstanceBlock] = {
+    require(maxMemoryUsage > 0)
+    val buff = mutable.ArrayBuilder.make[Instance]
+    var numCols = -1L
+    var count = 0L
+    var nnz = 0L
+    var allUnitWeight = true
+
+    iterator.flatMap { instance =>
+      if (numCols < 0L) numCols = instance.features.size
+      require(numCols == instance.features.size)
+      val n = instance.features.numNonzeros
+      var block = Option.empty[InstanceBlock]
+      // Check if enough memory remains to add this instance to the block.
+      if (getBlockSize(numCols, count + 1L, nnz + n,
+        allUnitWeight && (instance.weight == 1)) > maxMemoryUsage) {
+        // Check if this instance is too large
+        require(count > 0, s"instance $instance exceeds memory limit $maxMemoryUsage, " +
+          s"please increase block size")
+
+        block = Some(InstanceBlock.fromInstances(buff.result()))
+        buff.clear()
+        count = 0L
+        nnz = 0L
+        allUnitWeight = true
+      }
+      buff += instance
+      count += 1L
+      nnz += n
+      allUnitWeight &&= (instance.weight == 1)
+      block.iterator
+    } ++ {
+      val instances = buff.result()
+      if (instances.nonEmpty) {
+        Iterator.single(InstanceBlock.fromInstances(instances))
+      } else Iterator.empty
+    }
+  }
```

Comment:
This iterator logic here we'd better use for loop with `yield`, it will be more clear to read.